### PR TITLE
[Choice][Checkbox][RadioButton] Use CSS-native :hover styling

### PR DESCRIPTION
**Pull request by jesstelford**
_See https://github.com/Shopify/polaris/pull/9593 for context and why this is useful beyond just fixing a bug._

By using CSS-native hover styling, and pulling it up to the wrapping `<label>` it avoids an issue where the browser would briefly detect hover with the :hover pseudo selector setting `cursor: pointer`, which then got overwritten almost immediately with the JS `onMouseOver` styles to `cursor: default`, resulting in a flash of a pointy-hand for the cursor on disabled Checkbox.

https://github.com/Shopify/polaris/assets/612020/92a593b8-c670-46e9-ac28-790418574c07

> _Had to use my phone's "Slow Motion" camera to capture a video of it; none of the desktop recording software could get it_ 😓

- See the difference:
  - [This PR's Storybook](https://5d559397bae39100201eedc1-nonfodhlln.chromatic.com/?path=/story/all-components-checkbox--disabled-states&globals=polarisSummerEditions2023:true)
  - [`main`'s latest Storybook](https://5d559397bae39100201eedc1-oudurgatiw.chromatic.com/?path=/story/all-components-checkbox--disabled-states&globals=polarisSummerEditions2023:true)
- No visual regressions [according to Chromatic](https://shopify.chromatic.com/build?appId=5d559397bae39100201eedc1&number=18161).

### CODE REVIEW

File: polaris-react/src/components/Checkbox/tests/Checkbox.test.tsx [hunk 1]
```diff
@@ -251,31 +251,6 @@ describe('<Checkbox />', () => {
       });
     });
   });
-
-  describe('Hovering the label', () => {
-    it('adds the hover class to the Backdrop onMouseOver the label', () => {
-      const checkBox = mountWithApp(<Checkbox label="checkbox" />);
-
-      const label = checkBox.find('label');
-      label!.trigger('onMouseOver');
```
Comment (jesstelford) [inline polaris-react/src/components/Checkbox/tests/Checkbox.test.tsx:260]: This doesn't exist anymore.

... Also, why are our tests asserting against internal implementation details?

File: polaris-react/src/components/RadioButton/RadioButton.scss [hunk 2]
```diff
@@ -5,72 +5,109 @@
...
-      outline: var(--p-border-width-2) solid
-        var(--p-color-border-interactive-focus);
-      outline-offset: var(--p-space-025);
-    }
+    &,
+    // Ideally we'd use a single `:where(.ChoiceLabel:hover) &` after all these
+    // selectors to set the hover styles, but because of browser support, we
+    // have to ensure the specificity of these rules are maintained when hovered
+    // and not hovered.
+    .ChoiceLabel:hover & {
```
Comment (gwyneplaine) [inline polaris-react/src/components/RadioButton/RadioButton.scss]: We shouldn't need to add the .ChoiceLabel:hover specificity for focus-ring styles, can likely remove this.

File: polaris-react/src/components/Choice/Choice.tsx [hunk 3]
```diff
@@ -40,24 +38,18 @@ export function Choice({
   labelHidden,
   helpText,
   onClick,
-  onMouseOut,
```
Comment (gwyneplaine) [inline polaris-react/src/components/Choice/Choice.tsx:43]: Weary of removing onMouseOut and onMouseOver as props on `Choice` 
Considering there are [usages of this component ](https://grokt.shopify.io/results?q=%3CChoice+lang%3Atsx+-r%3APolaris) outside of Polaris.

**Review by gwyneplaine — CHANGES_REQUESTED** [review]

File: polaris-react/src/components/Choice/Choice.tsx (hunk 3 above)
Comment (jesstelford) [reply polaris-react/src/components/Choice/Choice.tsx:43]: Thankfully that's not Polaris's `Choice`; it's [not even exported from `@shopify/polaris`](https://github.com/Shopify/polaris/blob/main/polaris-react/src/index.ts), and the only usage within Polaris is RadioButton/Checkbox. 🎉

**Review by gwyneplaine — APPROVED** [review]

**Pull request merged**